spectrometer.py

```diff
@@ -1,303 +1,331 @@
-import os,sys, math, time
-from PIL import Image, ImageDraw, ImageFile, ImageFont
+import sys
+import math
+import time
 import picamera
-from collections import OrderedDict
 from fractions import Fraction
+from collections import OrderedDict
+from PIL import Image, ImageDraw, ImageFile, ImageFont
+
+
+# scan a column to determine top and bottom of area of lightness
+def get_spectrum_y_bound(pix, x, middle_y, spectrum_threshold, spectrum_threshold_duration):
+    c = 0
+    spectrum_top = middle_y
+    for y in range(middle_y, 0, -1):
+        r, g, b = pix[x, y]
+        brightness = r + g + b
+        if brightness < spectrum_threshold:
+            c = c + 1
+            if c > spectrum_threshold_duration:
+                break
+        else:
+            spectrum_top = y
+            c = 0
+
+    c = 0
+    spectrum_bottom = middle_y
+    for y in range(middle_y, middle_y * 2, 1):
+        r, g, b = pix[x, y]
+        brightness = r + g + b
+        if brightness < spectrum_threshold:
+            c = c + 1
+            if c > spectrum_threshold_duration:
+                break
+        else:
+            spectrum_bottom = y
+            c = 0
+
+    return spectrum_top, spectrum_bottom
+
+
+# find aperture on right hand side of image along middle line
+def find_aperture(pic_pixels, pic_width: int, pic_height: int) -> object:
+    middle_x = int(pic_width / 2)
+    middle_y = int(pic_height / 2)
+    aperture_brightest = 0
+    aperture_x = 0
+    for x in range(middle_x, pic_width, 1):
+        r, g, b = pic_pixels[x, middle_y]
+        brightness = r + g + b
+        if brightness > aperture_brightest:
+            aperture_brightest = brightness
+            aperture_x = x
+
+    aperture_threshold = aperture_brightest * 0.9
```

Comment:
General observation #2. So, values like `0.9` above are what are known as ["magic numbers"](https://en.wikipedia.org/wiki/Magic_number_(programming)) (see the section "Unnamed numerical constants.")

They're generally considered bad practice for a number of reasons: a) if you need to change one, you'll need to search through file(s) and possibly modify multiple lines/files for a single change, b) they provide no semantic value. Example:

```python
# what is 9.80665????
x = something * 9.80665
```

vs
```python
GRAVITATIONAL_CONSTANT = 9.80665
x = something * GRAVITATIONAL_CONSTANT
```

The latter instantly provides meaning and context to help understand what the code is doing. 

Now, the problem here is that there are *so* *many* magic constants; I wouldn't try to fix all of them at once. Better to strategically find ones that make the most sense and extract them to named constants. BTW, PyCharm provides a nice way to do this with option-cmd-V.

Reply:
Cool! Thanks for the info. I agree that the level of abstraction on a line needs to remain consistent. I'll work on fixing these up.